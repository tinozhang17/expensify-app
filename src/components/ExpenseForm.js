import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    defaultState = {
        description: this.props.expense ? this.props.expense.description : '',
        note:  this.props.note ? this.props.expense.description : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toFixed(2).toString() : '' ,
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false,
        error: undefined,
    };

    state = {
        ...this.defaultState,
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }))
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(?:\.\d{0,2})?$/)) {
            this.setState({ amount });
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: undefined }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
            this.setState((prevState) => ({...prevState}));
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange}/>
                    <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
                    <SingleDatePicker date={this.state.createdAt}
                                      onDateChange={this.onDateChange}
                                      focused={this.state.calendarFocused}
                                      onFocusChange={this.onFocusChange}
                                      numberOfMonths={1}
                                      isOutsideRange={() => false}/>
                    <textarea placeholder={"Add a note for your expense (optional)"} onChange={this.onNoteChange} />
                    <button>{this.props.expense ? 'Update Expense' : 'Add Expense'}</button>
                </form>
            </div>
        );
    }
}