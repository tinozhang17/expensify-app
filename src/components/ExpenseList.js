import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='content-container'>
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
        {
            props.expenses.length ? (props.expenses.map(expense => (<ExpenseListItem key={expense.id} dispatch={props.dispatch} {...expense} />))) : (<div className="list-item--message"><span>No Expenses</span></div>)
        }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList); // Now, ExpenseList gets access to the list of expenses in the state through the props argument. The connect() function basically creates a higher-order component