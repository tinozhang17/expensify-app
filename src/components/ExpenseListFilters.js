import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        startDate: null,
        endDte: null,
        calendarFocused: null,
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState({ calendarFocused });
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        const val = e.target.value;
        if (val === 'amount') {
            this.props.sortByAmount();
        } else if (val === 'date') {
            this.props.sortByDate();
        }
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input placeholder="Search expenses" className="text-input" type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select className="select" placeholder="Search expenses" onChange={this.onSortChange}>
                            {['Amount', 'Date'].map((item) => {
                                return (
                                    <option
                                        value={item.toLowerCase()}
                                        key={item.toLowerCase()}
                                        selected={item.toLowerCase()===this.props.filters.sortBy}>{item}</option>
                                );
                            })}
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={'filters__start-date-id'}
                            endDate={this.props.filters.endDate}
                            endDateId={'filters__end-date-id'}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    };
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (e) => dispatch(setTextFilter(e.target.value)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);