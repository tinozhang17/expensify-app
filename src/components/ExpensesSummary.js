import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeraljs';
import selectVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenses.length}</span> expense{props.expenses.length > 1 ? 's' : ''} totalling <span>{numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00')}</span></h1>
            <div>
                <div className="page-header__action">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    </div>
);

const mapStateToProps = (state) => ({
    expenses: selectVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);