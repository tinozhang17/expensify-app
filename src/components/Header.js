import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <button onClick={startLogout}>Log Out</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);