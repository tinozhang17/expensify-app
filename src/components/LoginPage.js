import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from "../actions/auth";

export const LoginPage = ({ startLogin }) => (
    <div>
        <form>
            <label htmlFor="login__username">Username: </label>
            <input id="login__username" type="text"/>
            <label htmlFor="login__password">Password: </label>
            <input id="login__password" type="text"/>
            <button type="button" onClick={startLogin}>Log In</button>
        </form>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);