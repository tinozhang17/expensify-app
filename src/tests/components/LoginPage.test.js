import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

let wrapper, startLoginSpy;

beforeEach(() => {
    startLoginSpy = jest.fn();
    wrapper = shallow(<LoginPage startLogin={startLoginSpy} />);
});

test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle login click correctly', () => {
    wrapper.find('button').simulate('click');
    expect(startLoginSpy).toHaveBeenCalled();
});