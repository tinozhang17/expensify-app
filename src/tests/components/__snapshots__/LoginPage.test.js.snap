// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <form>
    <label
      htmlFor="login__username"
    >
      Username: 
    </label>
    <input
      id="login__username"
      type="text"
    />
    <label
      htmlFor="login__password"
    >
      Password: 
    </label>
    <input
      id="login__password"
      type="text"
    />
    <button
      onClick={[MockFunction]}
      type="button"
    >
      Log In
    </button>
  </form>
</div>
`;
